### PR TITLE
NFT Changes

### LINKED ISSUE #828: NFT: restrict file access options.
Allocation changes to support NFTs.

1.  [x] Allocations cannot have the duration reduced. Change for all allocations.
4.  [x] Cannot be made mutable once set immutable.
3.  [x] Add an `anyone can extend` boolean flag.
5.  [x] Six new boolean options, `upload`,`delete`, `update`, `move`, `copy` and `rename`. These toggle whether the owner or anyone else can upload, delete, update, move, copy or rename files respectively.

This continues https://github.com/0chain/0chain/issues/538  Points (2) and (3) done in https://github.com/0chain/0chain/pull/620 https://github.com/0chain/gosdk/pull/288 and https://github.com/0chain/zboxcli/pull/122

Edit: Removed references to NFT, all CRUD operations default to true as below.

### PR DESCRIPTION
**Pull request by devyetii**
## Fixes
restrict file access options
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber: 0chain/blobber#891
- gosdk: 0chain/gosdk#687
- system_test: 0chain/system_test#495
- zboxcli: 0chain/zboxcli#335
- zwalletcli:
- actions: 0chain/actions#94
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 828-nft-restrict-file-access-options  |
| authorizer  | staging  |
| 0box  | 828-nft-restrict-file-access-options  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by devyetii** [comment]
Successful run here => https://github.com/0chain/system_test/actions/runs/3774572155/jobs/6416727470

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | staging  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 1]
```diff
@@ -1146,102 +1152,132 @@ func (sc *StorageSmartContract) updateAllocationRequestInternal(
...
-		}
-	}
 
-	if len(blobbers) != len(alloc.BlobberAllocs) {
-		return "", common.NewError("allocation_updating_failed",
-			"error allocation blobber size mismatch")
-	}
+	// If the txn client_id is not the owner of the allocation, should just be able to extend the allocation if permissible
+	// This way, even if an atttacker of an innocent user incorrectly tries to modify any other part of the allocation, it will not have any effect
+	if t.ClientID != alloc.Owner /* Third-party actions */ {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:1170]: is it possible to extend allocation without changing expiration?

Comment (devyetii) [reply code/go/0chain.net/smartcontract/storagesc/allocation.go:1170]: Afaik (and cmiiw) extending the allocation means either increasing size or extending expiration time. So yes the user (third-party) can either increase size alone, expiry alone or both. Are we on the same page regarding the meaning of extending the allocation ?

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Review by dabasov — APPROVED** [review]
lgtm

**Review by peterlimg — COMMENTED** [review]
Please resolve the conflicts

**Review by peterlimg — DISMISSED** [review]
Please add unit tests to confirm the changes work as expected. At least test:
- Allocation is not allowed to change if the `ThirdPartExtendable` is not set
- Allocation is only extendable when `ThirdPartExtendable` is set and the txn.Client is not owner
- Allocation owner is not affected by the `ThirdPartyExtenable` field, etc.

**Comment by devyetii** [comment]
@peterlimg Sorry I don't get that, which conflicts should I resolve ? The branch is up-to-date with staging I think.

**Comment by devyetii** [comment]
@peterlimg Added tests. Please review.

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | 686-add-fields-to-allocation-to-support-nft  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | 398-new-system-test-for-nft-allocation-features  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | 149-nft-changes  |
| zwalletcli  | staging  |
| 0chain  | 828-nft-restrict-file-access-options  |
| blobber  | 476-toggle-permission-to-uploaddelete-update-move-copy-and-rename  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by dabasov** [comment]
this PR is ready for merge, but we can break all the UI that doesn't support `FileOptions: 	 63,` and will create allocations that won't have any allowed actions

**Pull request merged**